server/keyshare/myirmaserver/server.go

```diff
@@ -428,174 +380,130 @@ func (s *Server) handleIrmaLogin(w http.ResponseWriter, r *http.Request) {
 		// not relevant for frontend, so ignore beyond log.
 	}
 
-	http.SetCookie(w, &http.Cookie{
-		Name:     "session",
-		Value:    session.token,
-		MaxAge:   s.conf.SessionLifetime,
-		Secure:   s.conf.Production,
-		Path:     "/",
-		HttpOnly: true,
-	})
+	s.setCookie(w, session.token, s.conf.SessionLifetime)
 	w.WriteHeader(http.StatusNoContent)
 }
 
+func (s *Server) logoutUser(w http.ResponseWriter, r *http.Request) {
+	session := s.sessionFromCookie(r)
+	if session != nil {
+		session.userID = nil // expire session
+	}
+	s.setCookie(w, "", -1)
+}
+
 func (s *Server) handleLogout(w http.ResponseWriter, r *http.Request) {
-	http.SetCookie(w, &http.Cookie{
-		Name:     "session",
-		Value:    "",
-		Secure:   s.conf.Production,
-		Path:     "/",
-		HttpOnly: true,
-	})
+	s.logoutUser(w, r)
 	w.WriteHeader(http.StatusNoContent)
 }
 
 func (s *Server) handleUserInfo(w http.ResponseWriter, r *http.Request) {
-	session := r.Context().Value("session").(*Sessiondata)
-	if session == nil || session.userID == nil {
-		s.conf.Logger.Info("Malformed request: not logged in")
-		server.WriteError(w, server.ErrorInvalidRequest, "Not logged in")
-		return
-	}
-
-	userinfo, err := s.db.UserInformation(*session.userID)
+	session := r.Context().Value("session").(*session)
+	user, err := s.db.user(*session.userID)
 	if err != nil {
 		s.conf.Logger.WithField("error", err).Error("Problem fetching user information from database")
 		server.WriteError(w, server.ErrorInternal, err.Error())
 		return
 	}
 
-	if userinfo.Emails == nil {
-		userinfo.Emails = []UserEmail{}
+	session.expiry = time.Now().Add(time.Duration(s.conf.SessionLifetime) * time.Second)
+	s.setCookie(w, session.token, s.conf.SessionLifetime)
+
+	if user.Emails == nil {
+		user.Emails = []userEmail{}
 	} // Ensure we never send nil in place of an empty list
-	server.WriteJson(w, userinfo)
+	server.WriteJson(w, user)
 }
 
 func (s *Server) handleGetLogs(w http.ResponseWriter, r *http.Request) {
 	offsetS := chi.URLParam(r, "offset")
 	offset, err := strconv.Atoi(offsetS)
 	if err != nil {
 		s.conf.Logger.WithField("error", err).Info("Malformed offset")
-		w.WriteHeader(http.StatusNotFound)
-		return
-	}
-
-	session := r.Context().Value("session").(*Sessiondata)
-	if session == nil || session.userID == nil {
-		s.conf.Logger.Info("Malformed request: user not logged in")
-		server.WriteError(w, server.ErrorInvalidRequest, "Not logged in")
+		server.WriteError(w, server.ErrorInvalidRequest, err.Error())
 		return
 	}
 
-	entries, err := s.db.Logs(*session.userID, offset, 11)
+	session := r.Context().Value("session").(*session)
+	entries, err := s.db.logs(*session.userID, offset, 11)
 	if err != nil {
 		s.conf.Logger.WithField("error", err).Error("Could not load log entries")
 		server.WriteError(w, server.ErrorInternal, err.Error())
 		return
 	}
 
+	session.expiry = time.Now().Add(time.Duration(s.conf.SessionLifetime) * time.Second)
+	s.setCookie(w, session.token, s.conf.SessionLifetime)
+
```

Comment:
Perhaps, but not I think in a way that is worth the decrease in code duplication:
* It cannot be done in `sessionMiddleware`, because that is also used by `/user/delete` which deletes the cookie. So we would have to create another middleware that is not used by that endpoints and is used by the others having those lines of code;
* Writing the cookie header, i.e. invoking `s.setCookie()`m has to happen before the handling of the request which writes the HTTP body, since the body may be written only after all headers are written. But writing the cookie at that time triggers Go's http code to emit HTTP status 200, causing the HTTP status that for example `/email/remove` emits to change from 204 to 200.

Reply:
Does it really change the status code? If I check the source code it only writes the cookie header to the response. My first impression would be that the status code wouldn't change by only doing that. That would be really weird behaviour if that actually occurs.

The first element about checking for `DELETE` doesn't have to be that difficult by the way, because you already have the request there. You can simply do something like:
```
if r.Method != http.MethodDelete {
  ...
}
```

Reply:
> Does it really change the status code? If I check the source code it only writes the cookie header to the response. My first impression would be that the status code wouldn't change by only doing that. That would be really weird behaviour if that actually occurs.

It really does, yes. I tried, and it breaks one of the unit tests that asserts that one endpoint returns 204. Using middleware that emits a cookie makes it emit 200 instead, breaking the test. It makes sense, as well, since the first line of an HTTP response must be the status code and only after that can the headers be printed.

As to checking in the middleware that the method is not `http.MethodDelete`: that would work, but I think the point of middleware is generally to apply *generic* functionality to a bunch of endpoints. Checking for the method breaks that principle, because it puts part of the logic of these endpoints inside this middleware. Of course one may deviate from such principles when the circumstances call for it, but combined with the previous point (the status codes), to me it doesn't seem worth it in this case.

Reply:
Weird that it behaves that way, but okay then it makes sense to do it in this way.